go/arrow/internal/testing/types/extension_types.go

```diff
@@ -18,20 +18,171 @@
 package types
 
 import (
+	"bytes"
 	"encoding/binary"
 	"fmt"
 	"reflect"
+	"strings"
+
+	"github.com/goccy/go-json"
 
 	"github.com/apache/arrow/go/v12/arrow"
 	"github.com/apache/arrow/go/v12/arrow/array"
+	"github.com/google/uuid"
 	"golang.org/x/xerrors"
 )
 
+type UUIDBuilder struct {
+	*array.ExtensionBuilder
+}
+
+func NewUUIDBuilder(bldr *array.ExtensionBuilder) *UUIDBuilder {
+	b := &UUIDBuilder{
+		ExtensionBuilder: bldr,
+	}
+	return b
+}
+
+func (b *UUIDBuilder) Append(v uuid.UUID) {
+	b.ExtensionBuilder.Builder.(*array.FixedSizeBinaryBuilder).Append(v[:])
+}
+
+func (b *UUIDBuilder) UnsafeAppend(v uuid.UUID) {
+	b.ExtensionBuilder.Builder.(*array.FixedSizeBinaryBuilder).UnsafeAppend(v[:])
+}
+
+func (b *UUIDBuilder) AppendValues(v []uuid.UUID, valid []bool) {
+	data := make([][]byte, len(v))
+	for i, v := range v {
+		data[i] = v[:]
+	}
+	b.ExtensionBuilder.Builder.(*array.FixedSizeBinaryBuilder).AppendValues(data, valid)
```

Comment:
this could be done more efficiently if desired via resize + `UnsafeAppend` or other methods. Not necessary for this PR but just something to think about.